core/services/workflows/syncer/backoff.go

```diff
@@ -0,0 +1,105 @@
+package syncer
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type backoffRecord struct {
+	retryCount  int
+	nextRetryAt time.Time
+	signature   string
+}
+
+type backoffStrategy struct {
+	tickInterval       time.Duration
+	maxBackoffInterval time.Duration
+	backoffRecords     map[string]backoffRecord
+	handler            func(ctx context.Context, event Event) error
+	lggr               logger.Logger
+}
+
+func newBackoffStrategy(lggr logger.Logger, handler func(ctx context.Context, event Event) error, tickInterval, maxBackoffInterval time.Duration) *backoffStrategy {
+	return &backoffStrategy{
+		tickInterval:       tickInterval,
+		maxBackoffInterval: maxBackoffInterval,
+		backoffRecords:     make(map[string]backoffRecord),
+		handler:            handler,
+		lggr:               lggr.Named("BackoffStrategy"),
+	}
+}
+
+func (b *backoffStrategy) Apply(ctx context.Context, event Event) {
+	key, sig, err := keysFor(event)
+	if err != nil {
+		b.lggr.Errorw("failed to apply backoff strategy: could not calculate keys for event", "err", err, "type", event.GetEventType())
+		return
+	}
+
+	// We have an existing backoffRecord for the key, but it might correspond to a different
+	// event. If it does, let's treat this as a reset and delete the record.
+	// If not, it's the same record, so we'll inspect it when determining whether to execute or not.
+	br, ok := b.backoffRecords[key]
+	if ok && br.signature != sig {
+		b.lggr.Debugw("found record matching workflowID with different signature, deleting...", "key", key, "oldSignature", br.signature, "newSignature", sig)
+		delete(b.backoffRecords, key)
+	}
+
+	// If we don't have a backoffRecord, or the time listed in the backoffRecord
+	// is in the past, we can execute now. Otherwise, log a message and continue.
+	br, ok = b.backoffRecords[key]
+	shouldExecuteNow := !ok || time.Now().After(br.nextRetryAt)
+	if !shouldExecuteNow {
+		b.lggr.Debugw("skipping event as it is still in backoff", "key", key, "nextRetryAt", br.nextRetryAt, "retryCount", br.retryCount)
+		return
+	}
+
+	err = b.handler(ctx, event)
+	if err != nil {
+		backoff := math.Pow(float64(2), float64(br.retryCount)) * float64(b.tickInterval)
```

Comment:
nit: maybe stop calling Pow() after a certain retryCount instead of comparing to maxBackoffInterval?